Setup Active Storage and Actionable Mailbox with Actionable Errors

gsamokovarov opened the pull request:
This is another take of #36043, that introduces actionable errors for the Active Storage and the Actionable Mailbox frameworks. The approach adds a hooking mechanism on the `ActionDispatch::ActionableExceptions` middleware that can reraise existing errors into actionable ones.

```ruby
# lib/action_mailbox/engine.rb
initializer "action_mailbox.setup" do
  ActionDispatch::ActionableExceptions.on ActiveRecord::StatementInvalid do |err|
    raise ActionMailbox::SetupError if err.to_s.match?(ActionMailbox::InboundEmail.table_name)
  end
end

module ActionMailbox
  # Generic base class for all Action Mailbox exceptions.
  class Error < StandardError; end

  # Raised when we detect that Action Mailbox has not been initialized.
  class SetupError < Error
    include ActiveSupport::ActionableError

    def initialize(message = nil)
      return super if message

      super("Action Mailbox uses a table in your application’s database named "
            "two and active_storage_attachments. To generate it run the following"
            "command:\n\n        rails action_mailbox:install")
    end

    action "Run action_mailbox:install" do
      Rails::Command.invoke "action_mailbox:install"
    end
  end
end
```

The API is quite generic and you have to be explicit about how to reraise the error.

gsamokovarov committed:
Introduce hooks support to ActionDispatch::ActionableExceptions

You can use the hooks to re-raise actionable errors from expected error
states.

gsamokovarov committed:
Setup Active Storage with an Actionable Error

gsamokovarov committed:
Setup Action Mailbox with an Actionable Error

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+  class Error < StandardError; end
+
+  # Raised when we detect that Action Mailbox has not been initialized.
+  class SetupError < Error
```
How about `MissingInstallError`? It's not like there was a problem in setting up the library.

gsamokovarov replied:
Yeah, I was just thinking about that. Maybe just `InstallError`?

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+require "active_support/actionable_error"
+require "rails/command"
```
Action Mailbox doesn't depend on Railties, so we can't just straight up require this here.

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+    include ActiveSupport::ActionableError
+
+    def initialize(message = nil)
+      return super if message
```
I find this early return difficult to understand. Why would there already be a message here?

Would it be clearer with?

```ruby
message ||= <<~MESSAGE
…
MESSAGE

super
```

gsamokovarov replied:
Of course!

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+      return super if message
+
+      super("Action Mailbox uses a table in your application’s database named " \
+            "two and active_storage_attachments. To generate it run the following" \
```
This exception message has an error in it. We don't have to mention the table names either, that's too low level for the abstraction we're making.

kaspth replied:
Does `action_mailbox:install` run migrations automatically? Otherwise we'd just raise the exception over and over.

gsamokovarov replied:
Yeah, we can improve the UX here. It currently creates a pending migration, which in turn issues another actionable error 😅

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+            "command:\n\n        rails action_mailbox:install")
+    end
+
+    action "Run action_mailbox:install" do
```
For the migration task we didn't say "Run db:migrate" but "Run migrations". What format should we follow for these action names?

gsamokovarov replied:
Hm, maybe a phrase will be better here. For example, for the setup of Action Mailbox to be actually complete, we may need to run Active Storage install command as well. The button name won't reflect the real action in that case, so no need to imply it. I'll change this. :+1:

kaspth commented on actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:
```diff
@@ -19,6 +27,9 @@ def call(env)
 
       redirect_to request.params[:location]
+    rescue Exception => err
+      hooks.each { |hook| hook.call(err) }
```
So the assumption is that hook will raise before we do below?

kaspth replied:
Let's just say `error` instead of `err` throughout. Plenty of room for the full word 😄

gsamokovarov replied:
> So the assumption is that hook will raise before we do below?

Yeah. I'm not sure how clever we have to be here. 🤷‍♂️

> Let's just say error instead of err throughout. Plenty of room for the full word.

My Go habits are leaking here. 😅

kaspth commented on actionmailbox/lib/action_mailbox/engine.rb:
```diff
@@ -29,5 +29,11 @@ class Engine < Rails::Engine
+
+    initializer "action_mailbox.setup" do
+      ActionDispatch::ActionableExceptions.on ActiveRecord::StatementInvalid do |err|
+        raise SetupError if err.to_s.match?(InboundEmail.table_name)
```
Should this use `error.message`?

gsamokovarov replied:
Yeah, I'm in the habit of calling `error.to_s` as setting a `@message` is a common mistake during the definition of custom exceptions, but I'll change that to clarify the examples.

kaspth commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,26 @@
+  class Error < StandardError; end
+
+  # Raised when we detect that Action Mailbox has not been initialized.
+  class SetupError < Error
```
Something seems odd to me about the abstraction here. The fact that we need to wrap this in a new exception and then call the message on that, seems needlessly roundabout. Don't really have a suggestion for now though.

gsamokovarov replied:
Maybe turn the API around "treat as actionable"? We can define the actions dynamically like:

```ruby
ActionDispatch::ActionableExceptions.on ActiveRecord::StatementInvalid, /#{InboundEmail.table_name}/ do |error| # Or anything responding to #=== for the check...
  # Treat this body as the class of an "actionable error". Implementation-wise,
  # we can monkey patch the current error object.
  action "Setup Action Mailbox" do
    Rails::Command.invoke("action_mailbox:install")
  end
end
```

gsamokovarov replied:
Can you check the current interface? I turned it around after the RailsConf feedback I received.

kaspth reviewed (commented):
We can talk about this at RailsConf too, if you'd rather wait 😄

gsamokovarov commented:
Yeah, let's discuss this at RailsConf. I'll start my traveling journey tomorrow and won't be available for a day or two anyway. I'll brainstorm on the API though. ✌️

gsamokovarov committed:
Introduce actionable errors triggers

The triggers let you reraise an existing error into actionable one.

gsamokovarov commented:
@kaspth, I actually took some of the feedback from RailsConf and introduced a new concept for the actionable errors called triggers. You can define triggers to raise the current actionable error if the trigger condition connects. You can check how the setup errors are implemented at the moment and give a bit of feedback. I actually like how it turned out, so let's see what you think.

There is still one problem, though. I can use `Rails::Command.invoke` because it calls `Kernel#exit` and this is troublesome if the process is not supposed to exit after the command invocation. It raises `SystemExit` exceptions which are actually caught by the error pages. 😅 Do you think we can make something like `Rails::Command.invoke` that forks the current process and spawns the command?

gsamokovarov commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,28 @@
+      MESSAGE
+    end
+
+    trigger ActiveRecord::StatementInvalid do |error|
```
This is the new interface. It says:

> Trigger the actionable error if an `ActiveRecord::StatementInvalid` is detected and it matches the condition given in the block, e.g. it's about the missing `action_mailbox_inbound_emails` table.

paracycle replied:
With this interface, though, the error class should not be autoloaded, since it can only register itself to the trigger list, if it was loaded. But if the error class is autoloaded, it will never get a chance to do that. Have you thought about a way around that?

gsamokovarov commented on actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:
```diff
@@ -19,6 +19,9 @@ def call(env)
 
       redirect_to request.params[:location]
+    rescue Exception => error
+      ActiveSupport::ActionableError.trigger_by(error)
```
The middleware tries to check whether the raised error can trigger an actionable error. Only one actionable error can be triggered at a time, as this just reraises an actionable error from the current one.

The nice part about 3rd party plugins defining actionable errors is that they don't need to care how to raise it if they define a trigger. The actionable exceptions middleware is in the stack anyway at this point.

gsamokovarov commented on actionmailbox/lib/action_mailbox/errors.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+
+    action "Install now" do
+      system "./bin/rails active_storage:install action_mailbox:install db:migrate"
```
I cannot use `Rails::Command.invoke` because it calls `Kernel#exit`.

paracycle commented:
> @kaspth, I actually took some of the feedback from RailsConf and introduced a new concept for the actionable errors called triggers. You can define triggers to raise the current actionable error if the trigger condition connects. You can check how the setup errors are implemented at the moment and give a bit of feedback. I actually like how it turned out, so let's see what you think.

I was meaning to come back to this PR and write down our conversation at Railsconf about this, for the record. I am really happy that you liked the approach. IMO it makes the whole API much nicer to work with.

Have fun at BalkanRuby! 👋

paracycle reviewed (commented):
This looks really solid!

itsWill commented on activesupport/CHANGELOG.md:
```diff
@@ -1,3 +1,33 @@
+        MESSAGE
+      end
+
+      trigger ActiveRecord::StatementInvalid do |error|
```
What do you think of an api similar to the active record callbacks like: 

``` ruby
trigger on: ActiveRecord::StatementInvalid, if: ->(error) { error.to_s.match?(InboundEmail.table_name) }
```

gsamokovarov replied:
Yes! This interface communicates the intent better.

gsamokovarov committed:
Clarify the intention of ActionError trigger class macro

Instead of:

```ruby
trigger ActiveRecord::StatementInvalid do |error|
  error.to_s.match?(InboundEmail.table_name)
end
```

Use the one below, as it makes the intention and direction of the
trigger clear:

```ruby
trigger on: ActiveRecord::StatementInvalid, if: -> error do
  error.to_s.match?(InboundEmail.table_name)
end
```

gsamokovarov committed:
Display the output of invoked actions

gsamokovarov commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -43,6 +55,20 @@ module ClassMethods
+      #       error.to_s.match?(InboundEmail.table_name)
+      #     end
+      #   end
+      def trigger(error, &condition)
```
I've changed the interface to:

```ruby
trigger on: AnError, if: -> it { matches_a_condition }
```

gsamokovarov commented:
Regarding https://github.com/rails/rails/pull/36071#discussion_r284831546, that's something that I cannot avoid with the current interface. Me and @kaspth tinkered about an interface that may live in the Railtie and solve the autoloading issue, but it still feels a bit off and hides the actionable errors a bit. It may have looked like:

```ruby
class ActiveStorage::Engine < Rails::Engine
  initializer "active_storage.actionable_errors" do
    ActiveSupport::ActionableError
      .action("Install now") { system "./bin/rails active_storage:install db:migrate" }
      .trigger(on: ActiveStorageSupport, if: -> error { table_name.in?(error.message) })
  end
end
```

I welcome feedback on ☝️ as it may be a possibility as well.

---

As for the current `triggers` concept, I have tweaked the interface and started collecting the output of the commands on the actionable error pages. They also don't feature a stack trace now. Here is a demo:


![ezgif-3-33959c2b620e](https://user-images.githubusercontent.com/604618/58583930-c3334f80-825d-11e9-9a0b-775eaeeaf4aa.gif)

gsamokovarov commented on actionpack/lib/action_dispatch/middleware/actionable_exceptions.rb:
```diff
@@ -16,24 +16,54 @@ def call(env)
+        end
+      end
+
+      def capture(stream)
```
We use this in testing... Capturing the streams may be tricky, but since most of the actions do dispatch commands, seems like the thing to do.

paracycle commented on activesupport/lib/active_support/actionable_error.rb:
```diff
@@ -43,6 +55,23 @@ module ClassMethods
+      #     end
+      #   end
+      def trigger(options)
+        error = options.fetch(:on) { raise ArgumentError, "missing keyword: on" }
```
Would it still not be more helpful to explicitly check that `error` is an `Exception` at invocation time, rather than allowing the action to never be triggered due to a bad constant that was passed for `on:`?

gsamokovarov replied:
I'm not sure how defensive we want to be here. The `trigger` method is used in the context of a class like:

```ruby
class InstallError
  trigger on: StandardError, if: condition
  # You probably won't be rescuing an error in the body of a class.
end
```

In the body of the class, you probably won't be catching already risen errors, but you'd have access to the exception classes themselves. Guarding for this case would make the code more verbose, while I don't currently see this as a common mistake to need the guard and make the check's worth, in a way. I can change the naming, though, to prevent any confusion here.

paracycle replied:
Sorry, I am not sure I follow your response and I suspect we could be talking about different things. Let me try to clarify what I mean. 

If I were to do something like this:
```ruby
class StupidError < StandardError
  include ActiveSupport::ActionableError

  trigger on: ActiveModel, if: condition
end
```
then this will never work since I did not pass a subclass of `Exception` or `StandardError` to the `on` parameter of the `trigger` method.

However, nothing will give me any indication that this will never work either as the PR currently stands. I am just suggesting that `trigger` validates that what is passed to `on:` is a subclass of `StandardError` and raises an error if it is not.

gsamokovarov replied:
If you want to break it, sure, you can! However, I'd prefer if we're not so defensive here.

Ruby's waters are quite muddy when it comes to what can rescue an exception. For example, the following compiles just fine:

```ruby
begin
  raise "Runtime Error"
rescue String
  p "Hey, I'm never called!"
end

#=> nil
```

Of course, it does nothing, but proves a point. 😅 Ruby's rescue doesn't match errors strictly based on their type hierarchy. It matches the errors with the case operator (`===`) and this let's you do interesting things like:

```ruby
def String.===(other)
  RuntimeError === other or super
end

begin
  raise "Runtime Error"
rescue String
  p "Look, Ma, I rescue String's (RuntimeError) now!"
end

#=> "Look, Ma, I rescue String's (RuntimeError) now!"
```

The only check `rescue` does to its rhs is that it's a module, not an `Exception` descendant per se. If its a module, it uses the `===` operator and you can have dynamic error catchers. They are surprisingly useful if you want to suppress noisy errors that don't adhere to a specific hierarchy, but you can still bring them together under one umbrella, or... it let's you rescue errors even based on their `#message` or other properties, if you inherit a messy codebase. _(See [masse](https://github.com/gsamokovarov/masse) for example.)_

If we want to be really correct in guarding the `trigger on:`  case, we'll have to write all those checks and I don't see any clear win for us here. If it was something that's called often and would have blown up mystically later on, maybe I would have gone the extra mile and guard against such cases. However, the way I see the actionable errors trigger usage at the moment is as follows.

You define it once, you test that the error is triggered and... everything is kinda static after that. The triggers (and the actionable errors) are not dynamically defined and probably won't be used that much. So, if you get the error to trigger once, you're good. You have to test the actual trigger code more, not just the error you passed to it.

I also want to keep the door open for dynamic error matcher support* and  the following interface:

```ruby
trigger on: 'ActiveRecord::StatementInvalid", if: -> error do
  # ...
end
```

This could be useful if we want to delay the autoloading of `ActiveRecord::StatementInvalid` should the current actionable errors load Active Record too soon.

_* Currently, the triggers doesn't match the errors with `#===` for simplicitly, but if someone needs it, we can support it later on._

paracycle replied:
Ok, convinced! 😄 Thank you for that extensive and well-written explanation, makes it crystal clear why such a check might be unnecessary! 👍

paracycle commented:
@gsamokovarov I personally like the self contained error classes as they are defined now in the PR as opposed to declaring the trigger/action in an initializer. The initializer also makes it harder to override the `message` as well.

An idea I have to solve the autoloading would be to add a `ActiveSupport::ActionableError.register` method that takes the actionable error type as an argument. As of now, there isn't much that method can do, but if we make it mandatory that `register` is called in an initializer before an actionable error starts working, we can ensure that actionable errors are autoloaded.

For example:
```ruby
class ActiveStorage::Engine < Rails::Engine
  initializer "active_storage.actionable_errors" do
    ActiveSupport::ActionableError.register(ActionMailbox::SetupError)
  end
end
```
and the implementation of `ActiveSupport::ActionableError.register` could be just:
```ruby
  module ActionableError
...
    mattr_accessor :triggers, default: {} # :nodoc:
...
    def self.register(error)
      triggers[error] = []
    end
...
    module ClassMethods
      def trigger(options)
        error = options.fetch(:on) { raise ArgumentError, "missing keyword: on" }
        condition = options.fetch(:if) { raise ArgumentError, "missing keyword: if" }

        raise SomethingError, "#{error} not registered" unless ActiveSupport::ActionableError.triggers[error]

        ActiveSupport::ActionableError.triggers[error] << Trigger.new(self, condition)
      end
    end
  end
```
i.e. we only add to `triggers` if it already has an error registered to it.

WDYT about that?

gsamokovarov committed:
Call the Active Storage actionable error InstallError

gsamokovarov committed:
Call the Action Mailbox actionable error InstallError

gsamokovarov committed:
Clarify what trigger_by means

gsamokovarov committed:
Mention the possible autoloading issues while triggering an error

gsamokovarov commented:
Hey Ufuk, thank you so much for the review! I was looking for you GitHub handle when I introduced the triggers and now I have it! 🙌

One of the reasons I like the idea of the triggers was that it's way easier to define an actionable error now. You don't need to care how to raise it and you can describe the flow in one place. Same with @kaspth's idea of maybe defining them in the railtie. Again, the whole definition is in one place.

If we try to work around the autoloading issues, we'll be splitting the work again in two places and this will, IMO. introduce needless friction. The actionable errors will be raised kinda late into the autoloading game and most of the app can be loaded at this point. I think we can solve that with documentation (added a not on the `trigger` class macro documentation), instead of more requiring more work from the developers on the "setup for actionable errors" side.

Also, autoloading is something the Ruby core team doesn't really like and I haven't seen it been used that much outside of Rails, where we use it as a speed-up hack, so I really hope that this won't be a problem for most of the 3rd parties, as long as it's documented.

Thanks again for the review! 🙇‍♂️

gsamokovarov committed:
Use Exception#message instead of #to_s

paracycle commented:
> Hey Ufuk, thank you so much for the review! I was looking for you GitHub handle when I introduced the triggers and now I have it! 🙌

Lol, no worries. I was late finding the PR and commenting on it myself. And by the time I'd done that, you'd already refactored. Love the final state of it. ❤️ 

> If we try to work around the autoloading issues, we'll be splitting the work again in two places and this will, IMO. introduce needless friction. 

Totally agreed. I was just raising it as a potential pitfall. We don't have to work around it.

> I think we can solve that with documentation (added a not on the `trigger` class macro documentation), instead of more requiring more work from the developers on the "setup for actionable errors" side.

Agreed on this point too.

> Thanks again for the review! 🙇‍♂️

You are most welcome. Glad if I could be of some help.

gsamokovarov committed:
Handle errors in the actions themselves

gsamokovarov committed:
Migrate to Rails::Command.invoke now that the actionable errors have own page

gsamokovarov committed:
Only define the actionable errors if run in Rails

gsamokovarov committed:
Setup acitonable errors for Action Text

gsamokovarov committed:
Follow the cause of the error if triggering an actionable error

In the Action Text case, the `ActiveRecord::StatementInvalid` error is
re-raised as an `ActionView::Template::Error`. Following the
`Exception#cause` let us handle such cases without listening for the
triggers on multiple errors.

gsamokovarov committed:
Requite the Action Mailbox errors only in rails

gsamokovarov commented:
This PR changed directions a few times already and it's getting hard to review. I have opened https://github.com/rails/rails/pull/36455 with a cleaner history and a better PR description. @paracycle, @itwill, @prathamesh-sonpatki and @kaspth, if you want to take another look, check-out ☝️
